news/_posts/2019-12-25-prestashop-in-2019-and-beyond-part-3-the-future-architecture.md

```diff
@@ -0,0 +1,269 @@
+
+#### Data-centric design
+
+PrestaShop was designed in a traditional data-centric way. In this kind of design, systems are built "from the ground up": business subjects (like Users, Products, etc) are modeled as simple objects, usually following the _database schema_, and then their behavior is given by one or many abstraction layers placed on top of that model. This design encourages thinking about every action in terms of data (as opposed to _state_), because in the end, all high level services are actually built on top of [CRUD actions](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) performed on thin, behaviorless models that are placed at the bottom and support the whole structure.
+
+![Data-centric design][data-centric-design]
+
+The problems associated with this architecture have been explained in previous article: upper layers are tightly coupled to models, because they are _based_ on thin models which by definition (since they have no behavior) **need to be micro-managed**. This makes iterative enhancements harder to implement because since each layer either micro-manages its lower layer or is closely dependent on its implementation, **any change in an upper layer usually requires a change in lower layers as well** so that it provides the "buttons and levers" that the upper layer needs. 
+
+As a consequence, this kind of architecture is more often than not coupled with a never ending stream of _"oops, this design didn't account for that special case"_ kind of issues, that pushes developers to create ad-hoc solutions, further aggravating the situation because of crisscrossing dependencies, responsibility dilution and increased complexity–all resulting in a codebase that is harder to maintain and more bug-prone. 
+
+All this is not new. As systems evolve and grow bigger, one of the hardest problems in software engineering is how to handle this inevitably increasing complexity. Eric Evans's [Domain-Driven Design](https://en.wikipedia.org/wiki/Domain-driven_design) ("DDD" for short), a software design approach that has been gaining popularity recently, attempts to tackle this problem by choosing a dramatically different angle: it advocates for focusing on the _core domain_ (ie. "business logic"–what the application _does_) and _ubiquitous language_ (ie. "business language", or _expressiveness_). 
+
+In my interpretation, this design strategy is opposed to data-centric designs (where _models_ reflect database elements and _services_ manipulate their data), because it focuses on modeling _business objects_ and its interactions, based on user's intent or that make sense from a business, real-world-case scenarii, and that respond to state transitions which also make sense from that standpoint. 
```

Comment:
```suggestion
In my interpretation, this design strategy is opposed to data-centric designs (where _models_ reflect database elements and _services_ manipulate their data), because it focuses on modeling _business objects_ and their interactions, based on user's intent or what make sense from a business, real-world-case scenarii, and that respond to state transitions which also make sense from that standpoint. 
```
I'm not sure about my suggestion 🤔and the sentence looks a bit too long, maybe ?

Reply:
How about this?

```suggestion
In my interpretation, this design strategy is opposed to data-centric designs, where _models_ reflect database elements and _services_ manipulate their data. Instead, DDD focuses on modeling _business objects_ and their interactions, based on user's intent or on what make sense from a business, real-world-case scenarii; it gives models behavior that responds to state transitions that make sense from a business standpoint.
```